include/exadg/incompressible_navier_stokes/time_integration/time_int_bdf.h

```diff
@@ -159,6 +219,10 @@ class TimeIntBDF : public TimeIntBDFBase<Number>
   // postprocessor
   std::shared_ptr<PostProcessorInterface<Number>> postprocessor;
 
+  // time history
+  double              time_history_np;
```

Comment:
note that I adjusted the above pseudo-code.

Reply:
Computing `get_quantities_and_times_np()` before do_timestep solve will pass the test and give wrong results (see below). In my opinion, the user should be able to obtain at least a vector of quantities and corresponding correct times, no matter when the function is called. 
```
[q^end, q^n, q^n-1, ..., q^end-1]
[t^n+1,  t^n,  t^n-1, ...,  t^end-1]
```

Reply:
I see it a bit differently. The function name `_np()` indicates that this function accesses post-solve information. There is no guarantee to have correct results before calculating the time step. Before solving, you can only call `get_quantities_and_times()` and expect correct results. We can also rename the function

`get_quantities_and_times_np() -> get_quantities_and_times_post_solve()`,

but it will remain a getter function, i.e. extracting information from a time integrator class that is in a certain state currently. I would argue time integration implementations generally suffer from the problem that you describe. However, I do not see stringent solutions that do not rely on any form of conventions.

Reply:
So you are saying exposing `BDFCoefficients` is not OK because it can be misused, but it is OK to have a function which might give information that is not just the information from the old time-step (which I would expect) but simply wrong?

Reply:
Ensuring that `solutions` and `times` are in sync is in my opinion clearer from a user perspective (and easier from an implementation perspective) than `solutions`, `times`, and `bdf_constants`. The reason is that `bdf_constants` are updated in `do_timestep_pre_solve()` 
https://github.com/exadg/exadg/blob/290dd00a2fb1a4cec698eae55ec3680379167262/include/exadg/time_integration/time_int_bdf_base.cpp#L208 

while `solutions` and `times` are updated in `do_timestep_post_solve()` 
https://github.com/exadg/exadg/blob/290dd00a2fb1a4cec698eae55ec3680379167262/include/exadg/time_integration/time_int_bdf_base.cpp#L217

Moreover, the function `get_solutions_and_times()` returns both solutions and times, while `get_bdf_constants()` would be an independent function for which it is somewhat difficult to ensure that this is consistent with `get_solutions_and_times()`.

Finally, the BDF coefficients live in the IncNS time integrator to compute the **velocity** time derivative. There is per se no guarantee in my opinion that the vector of pressure solutions fits to such coefficients (e.g. regarding the size of the vector / order of the finite difference approximation of the derivative). This is another argument to only return solution vectors and times (and compute the time derivative based on this data only, as suggested here https://github.com/exadg/exadg/pull/458#discussion_r1219424100). Mathematically, the BDF constants are redundant information in addition to the `times` vector. I think one can safely argue that this is indeed more dangerous than providing minimal / unique information.

Regarding your question: Yes, using a function not in a way as described by the name and the documentation is principally "dangerous", where "dangerous" means it does not do what it is described to do. One cannot expect a function with name `_np()` to provide correct data at time `np`, if `np` data has not been computed.

Overall, I feel the discussion got heated and therefore suggest a pause for now.

Reply:
@nfehn I still think the data returned should be at least in sync, but let's proceed as you described and discuss this at a different point. The implemented functionality works, as long as times and solutions are in sync (independent who cares for it). Just let me add a test because the functionality is currently not used in the code. I will ping you when I am ready.